bee-message/src/message.rs

```diff
@@ -0,0 +1,275 @@
+    pub fn verify(&self) -> Result<(), ValidationError> {
+        let ed25519_public_key = ed25519::PublicKey::from_compressed_bytes(self.issuer_public_key)?;
+
+        // Unwrapping is okay here, since the length of the signature is already known to be correct.
+        let ed25519_signature = ed25519::Signature::from_bytes(self.signature.to_vec().try_into().unwrap());
+
+        let hash = self.hash();
+
+        if !ed25519_public_key.verify(&ed25519_signature, &hash) {
+            Err(ValidationError::InvalidSignature)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Packable for Message {
+    type PackError = MessagePackError;
+    type UnpackError = MessageUnpackError;
+
+    fn packed_len(&self) -> usize {
+        self.parents.packed_len()
+            + self.issuer_public_key.packed_len()
+            + self.issue_timestamp.packed_len()
+            + self.sequence_number.packed_len()
+            + self.payload.packed_len()
+            + self.nonce.packed_len()
+            + self.signature.packed_len()
+    }
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), PackError<Self::PackError, P::Error>> {
+        self.parents.pack(packer)?;
+        self.issuer_public_key.pack(packer).map_err(PackError::infallible)?;
+        self.issue_timestamp.pack(packer).map_err(PackError::infallible)?;
+        self.sequence_number.pack(packer).map_err(PackError::infallible)?;
+        self.payload.pack(packer)?;
+        self.nonce.pack(packer).map_err(PackError::infallible)?;
+
+        let sig_bytes: [u8; MESSAGE_SIGNATURE_LENGTH] = self.signature.to_vec().try_into().unwrap();
+        sig_bytes.pack(packer).map_err(PackError::infallible)?;
+
+        Ok(())
+    }
+
+    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let parents = Parents::unpack(unpacker)?;
+        let issuer_public_key = <[u8; MESSAGE_PUBLIC_KEY_LENGTH]>::unpack(unpacker).map_err(UnpackError::infallible)?;
+        let issue_timestamp = u64::unpack(unpacker).map_err(UnpackError::infallible)?;
+        let sequence_number = u32::unpack(unpacker).map_err(UnpackError::infallible)?;
+        let payload = Option::<Payload>::unpack(unpacker).map_err(UnpackError::coerce)?;
+        let nonce = u64::unpack(unpacker).map_err(UnpackError::infallible)?;
+        let signature = <[u8; MESSAGE_SIGNATURE_LENGTH]>::unpack(unpacker)
+            .map_err(UnpackError::infallible)?
+            .into();
+
+        let message = Self {
+            parents,
+            issuer_public_key,
+            issue_timestamp,
+            sequence_number,
+            payload,
+            nonce,
+            signature,
+        };
+
+        let len = message.pack_to_vec().unwrap().len();
+        validate_message_len(len).map_err(|e| UnpackError::Packable(e.into()))?;
+
+        Ok(message)
+    }
+}
+
+/// A builder to build a `Message`.
+#[derive(Default)]
+pub struct MessageBuilder {
```

Comment:
Can you split the builder into its own module ?

Reply:
Sure, in it's own file?

Reply:
Yes